File: .github/actions/images/src/azure.ts
import * as core from '@actions/core';
import * as exec from '@actions/exec';
import { Image } from './types';

const RESOURCE_NOT_FOUND = 'Code: ResourceNotFound';
const SECRET_NOT_FOUND = '(SecretNotFound)';
const SECRET_FORBIDDEN = 'Caller is not authorized to perform action on resource.';

export async function getKeyVaultSecret(id: string): Promise<string | undefined> {

    try {

        const secretShowCmd = [
            'keyvault',
            'secret',
            'show',
            '--id', id,
        ];

        core.info(`Getting secret from Key Vault: ${id}`);
        const secretShow = await exec.getExecOutput('az', secretShowCmd, { silent: true, ignoreReturnCode: true });

        if (secretShow.exitCode === 0) {

            const secret = JSON.parse(secretShow.stdout);
            core.setSecret(secret.value);
            return secret.value;

        } else if (secretShow.stderr.includes(SECRET_NOT_FOUND))
            core.setFailed(`Key Vault secret ${id} not found`);
        else if (secretShow.stderr.includes(SECRET_FORBIDDEN))
            core.setFailed(`Service Principal used to login to Azure does not have permission to read Key Vault secret ${id}`);
        else
            core.setFailed(`Error getting secret from Key Vault: ${secretShow.stderr}`);

    } catch (error) {
        if (error instanceof Error) core.setFailed(error.message);
    }

    return undefined;
};

export async function validateImageDefinitionAndVersion(image: Image): Promise<boolean> {

    core.startGroup(`Validating image definition and version for ${image.name}`);

    try {
        const imgDefShowCmd = [
            'sig', 'image-definition', 'show',
            '--only-show-errors',
            '-g', image.galleryResourceGroup,
            '-r', image.galleryName,
            '-i', image.name
        ];

        core.info(`Checking if image definition exists for ${image.name}`);
        const imgDefShow = await exec.getExecOutput('az', imgDefShowCmd, { silent: true, ignoreReturnCode: true });

        if (imgDefShow.exitCode === 0) {

            core.info(`Found existing image definition for ${image.name}`);
            const img = JSON.parse(imgDefShow.stdout);
            image.location = image.useBuildGroup ? '' : img.location;

            // image definition exists, check if the version already exists

            const imgVersionShowCmd = [
                'sig', 'image-version', 'show',
                '--only-show-errors',
                '-g', image.galleryResourceGroup,
                '-r', image.galleryName,
                '-i', image.name,
                '-e', image.version
            ];

            core.info(`Checking if image version ${image.version} already exists for ${image.name}`);
            const imgVersionShow = await exec.getExecOutput('az', imgVersionShowCmd, { silent: true, ignoreReturnCode: true });

            if (imgVersionShow.exitCode !== 0) {
                if (imgVersionShow.stderr.includes(RESOURCE_NOT_FOUND)) {

                    // image version does not exist, add it to the list of images to create
                    core.info(`Image version ${image.version} does not exist for ${image.name}. Creating`);
                    // include.push(image);

                    core.endGroup();
                    return true;

                } else {

                    // image version exists, but we don't know what happened
                    core.setFailed(`Failed to check for existing image version ${image.version} for ${image.name} \n ${imgVersionShow.stderr}`);
                }

            } else if (image.changed) {

                // image version already exists, but the image source has changed, so we the version number needs to be updated
                core.setFailed(`Image version ${image.version} already exists for ${image.name} but image definition files changed. Please update the version number or delete the image version and try again.`);
            } else {

                // image version already exists, and the image source has not changed, so we don't need to do anything
                core.info(`Image version ${image.version} already exists for ${image.name} and image definition is unchanged. Skipping`);
            }

        } else if (imgDefShow.stderr.includes(RESOURCE_NOT_FOUND)) {

            // image definition does not exist, create it and skip the version check

            core.info(`Image definition for ${image.name} does not exist in gallery ${image.galleryName}`);

            const imgDefCreateCmd = [
                'sig', 'image-definition', 'create',
                '--only-show-errors',
                '-g', image.galleryResourceGroup,
                '-r', image.galleryName,
                '-i', image.name,
                '-p', image.publisher,
                '-f', image.offer,
                '-s', image.sku,
                '--os-type', image.os,
                // '--os-state', 'Generalized', (default)
                '--description', image.description,
                '--hyper-v-generation', 'V2',
                '--features', 'SecurityType=TrustedLaunch'
            ];

            core.info(`Creating new image definition for ${image.name}`);
            const imgDefCreate = await exec.getExecOutput('az', imgDefCreateCmd, { silent: true, ignoreReturnCode: true });

            if (imgDefCreate.exitCode === 0) {

                core.info(`Successfully created image definition for ${image.name}`);
                const img = JSON.parse(imgDefCreate.stdout);
                image.location = image.useBuildGroup ? '' : img.location;

                core.endGroup();
                return true;

            } else {
                core.setFailed(`Failed to create image definition for ${image.name} \n ${imgDefCreate.stderr}`);
            }

        } else {
            core.setFailed(`Failed to get image definition for ${image.name} \n ${imgDefShow.stderr}`);
        }

    } catch (error) {
        if (error instanceof Error) core.setFailed(error.message);
    }

    core.endGroup();
    return false;
};

File: builder/azure.py
# ------------------------------------
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
# ------------------------------------

import asyncio
import json
import os
import shutil
import subprocess
import sys
from pathlib import Path

import loggers

IMAGE_PARAMS_FILE = 'image.parameters.json'
RESOURCE_NOT_FOUND = 'Code: ResourceNotFound'
DEFAULT_PARAMS = ['name', 'location', 'version', 'tempResourceGroup', 'buildResourceGroup', 'gallery', 'replicaLocations']

log = loggers.getLogger(__name__)


def error_exit(message):
    log.error(message)
    sys.exit(message)


def _group_create_cmd(group_name, image):
    return ['group', 'create', '-n', group_name, '-l', image['location'], '--subscription', image['subscription']]


def _deployment_group_create_cmd(group_name, template_file, params_file, image):
    return ['deployment', 'group', 'create', '-n', image['name'], '-g', group_name, '-f', template_file,
            '-p', f'@{params_file}', '--no-prompt', '--subscription', image['subscription']]


def _img_def_show_cmd(image):
    return ['sig', 'image-definition', 'show', '--only-show-errors', '-g', image['gallery']['resourceGroup'],
            '-r', image['gallery']['name'], '-i', image['name'], '--subscription', image['gallery']['subscription']]


def _img_ver_show_cmd(image):
    return ['sig', 'image-version', 'show', '--only-show-errors', '-g', image['gallery']['resourceGroup'],
            '-r', image['gallery']['name'], '-i', image['name'], '-e', image['version'], '--subscription', image['gallery']['subscription']]


def _img_def_create_cmd(image):
    return ['sig', 'image-definition', 'create', '--only-show-errors', '-g', image['gallery']['resourceGroup'],
            '-r', image['gallery']['name'], '-i', image['name'], '-p', image['publisher'], '-f', image['offer'],
            '-s', image['sku'], '--os-type', image['os'], '--description', image['description'],
            '--hyper-v-generation', 'V2', '--features', 'SecurityType=TrustedLaunch', '--subscription', image['gallery']['subscription']]


def _parse_command(command):
    '''Parses a command (string or list of args), adds the required arguments, and replaces executable with full path'''
    if isinstance(command, list):
        args = command
    elif isinstance(command, str):
        args = command.split()
    else:
        raise ValueError(f'az command must be a string or list, not {type(command)}')

    az = shutil.which('az')

    if args[0] == 'az':
        args.pop(0)

    if args[0] != az:
        args = [az] + args

    return args


def cli(command, log_command=True):
    '''Runs an azure cli command and returns the json response'''
    args = _parse_command(command)

    try:
        if log_command:
            log.info(f'Running az cli command: {" ".join(args)}')

        proc = subprocess.run(args, capture_output=True, check=True, text=True)

        if proc.returncode == 0 and not proc.stdout:
            return None
        for line in proc.stdout.splitlines():
            log.info(line)

        resource = json.loads(proc.stdout)
        return resource

    except subprocess.CalledProcessError as e:

        if e.stderr and RESOURCE_NOT_FOUND in e.stderr:
            return None

        error_exit(e.stderr if e.stderr else 'azure cli command failed')

    except json.decoder.JSONDecodeError:
        error_exit('{}: {}'.format('Could not decode response json', proc.stderr if proc.stderr else proc.stdout if proc.stdout else proc))


def get_sub():
    '''Returns the current subscription id from the azure cli'''
    sub = cli('az account show')
    return sub['id']


def ensure_image_def_version(image):
    '''Ensures that the image definition exists and the version does not exist in the gallery'''
    image_name = image['name']
    image_version = image['version']

    build = False

    log.info(f'Validating image definition and version for {image_name}')
    log.info(f'Checking if image definition exists for {image_name}')
    imgdef = cli(_img_def_show_cmd(image))

    if imgdef:  # image definition exists, check if the version already exists

        log.info(f'Found existing image definition for {image_name}')
        log.info(f'Checking if image version {image_version} exists for {image_name}')
        imgver = cli(_img_ver_show_cmd(image))

        if imgver:
            log.info(f'Found existing image version {image_version} for {image_name}')
            log.warning(f'{image_name} was not built because version {image_version} already exists. Please update the version number or delete the existing image version and try again.')
        else:  # image version does not exist, add it to the list of images to create
            log.info(f'Image version {image_version} does not exist for {image_name}')
            build = True

    else:  # image definition does not exist, create it and skip the version check

        log.info(f'Image definition does not exist for {image_name}')
        log.info(f'Creating image definition for {image_name}')
        imgdef = cli(_img_def_create_cmd(image))

        build = True

    return build, imgdef


def save_params_file(image, params, filename):
    params_json = {
        '$schema': 'https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#',
        'contentVersion': '1.0.0.0',
        'parameters': {
            'image': {
                'value': image['name']
            },
            'version': {
                'value': image['version']
            }
        }
    }

    for p in params:
        params_json['parameters'][p] = {
            'value': params[p]
        }

    with open(Path(image['path']) / filename, 'w') as f:
        json.dump(params_json, f, ensure_ascii=False, indent=4, sort_keys=True)

    return os.path.join(image['path'], filename)


def deploy_builder(image, params_file):
    '''Deploys the builder resources to kick off the image build'''
    bicep_file = os.path.join(Path(__file__).resolve().parent, 'templates', 'builder.bicep')

    if 'tempResourceGroup' in image and image['tempResourceGroup']:
        group_name = image['tempResourceGroup']
        group = cli(_group_create_cmd(group_name, image))
    else:
        group_name = image['buildResourceGroup']

    dep = cli(_deployment_group_create_cmd(group_name, bicep_file, params_file, image))

    return dep


# ----------------
# async functions
# ----------------


async def cli_async(command, log_command=True):
    '''Runs an azure cli command and returns the json response'''
    args = _parse_command(command)

    if log_command:
        log.info(f'Running az cli command: {" ".join(args)}')

    try:
        proc = await asyncio.create_subprocess_exec(*args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
        stdout, stderr = await proc.communicate()

        if stderr and RESOURCE_NOT_FOUND in stderr.decode():
            return None

        if proc.returncode != 0:
            error_exit(stderr.decode() if stderr else 'azure cli command failed')

        if stdout:
            for line in stdout.decode().splitlines():
                log.info(line)

            try:
                resource = json.loads(stdout)
                return resource
            except json.decoder.JSONDecodeError:
                error_exit('{}: {}'.format('Could not decode response json', stderr.decode() if stderr else stdout if stdout else proc))

    except SystemExit as e:

        error_exit(e.code if e.code else 'azure cli command failed')


async def get_sub_async():
    '''Returns the current subscription id from the azure cli'''
    sub = await cli_async('az account show')
    return sub['id']


async def ensure_image_def_version_async(image):
    '''Ensures that the image definition exists and the version does not exist in the gallery'''
    image_name = image['name']
    image_version = image['version']

    build = False

    log.info(f'Validating image definition and version for {image_name}')
    log.info(f'Checking if image definition exists for {image_name}')
    imgdef = await cli_async(_img_def_show_cmd(image))

    if imgdef:  # image definition exists, check if the version already exists

        log.info(f'Found existing image definition for {image_name}')
        log.info(f'Checking if image version {image_version} exists for {image_name}')
        imgver = await cli_async(_img_ver_show_cmd(image))

        if imgver:
            log.info(f'Found existing image version {image_version} for {image_name}')
            log.warning(f'{image_name} was not built because version {image_version} already exists. Please update the version number or delete the existing image version and try again.')
        else:  # image version does not exist, add it to the list of images to create
            log.info(f'Image version {image_version} does not exist for {image_name}')
            build = True

    else:  # image definition does not exist, create it and skip the version check

        log.info(f'Image definition does not exist for {image_name}')
        log.info(f'Creating image definition for {image_name}')
        imgdef = await cli_async(_img_def_create_cmd(image))

        build = True

    return build, imgdef


async def deploy_builder_async(image, params_file):
    '''Deploys the builder resources to kick off the image build'''
    bicep_file = os.path.join(Path(__file__).resolve().parent, 'templates', 'builder.bicep')

    if 'tempResourceGroup' in image and image['tempResourceGroup']:
        group_name = image['tempResourceGroup']
        group = await cli_async(_group_create_cmd(group_name, image))
    else:
        group_name = image['buildResourceGroup']

    dep = await cli_async(_deployment_group_create_cmd(group_name, bicep_file, params_file, image))

    return dep
